### DISCUSSION TITLE
Crate stabilization

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2160] [General]
A post to consolidate all discussion, links, PRs etc. for crate stabalization. There is an issue on each crate for crate specific discussion. This is for things that apply to stabalizing each crate or the whole process.

Tracking issue for each crate:

- Current work in progress:
  - encoding: https://github.com/rust-bitcoin/rust-bitcoin/issues/5686 
  - units: #3556
  - hashes: #3554
  - io: #3555
  - hex: https://github.com/rust-bitcoin/hex-conservative/issues/43
  - ordered: https://github.com/rust-bitcoin/rust-ordered/issues/15
  - primitives: #3133
- Future work
  - bech32: https://github.com/rust-bitcoin/rust-bech32/issues/157
  - secp256k1: https://github.com/rust-bitcoin/rust-secp256k1/issues/662
  - bitcoin: https://github.com/rust-bitcoin/rust-bitcoin/issues/2158

### Potential issues that apply to all crates:

- https://github.com/rust-bitcoin/rust-bitcoin/issues/1649
- https://github.com/rust-bitcoin/rust-bitcoin/issues/1491
- https://github.com/rust-bitcoin/rust-bitcoin/issues/820
- https://github.com/rust-bitcoin/rust-bitcoin/issues/1780
- https://github.com/rust-bitcoin/rust-bitcoin/issues/1251
- #3826

### Related links

- https://github.com/rust-bitcoin/rust-bitcoin/discussions/1128
- https://github.com/rust-bitcoin/rust-bitcoin/discussions/1392

### Things we did in `units` that probably want doing in all crates 

- [ ] Enable pedantic lints (see #3704)
- [ ] Add an `api` integration test (see `units/tests/api.rs`)
- [ ] Add `must_use` (see #3715)
- [ ] Use `return_self_must_not_use` (see #3713)
- [ ] Implement `core::iter::Sum` for types that implement `Add` (see #3677)
- [ ] Close all the errors (e.g. #3647)

**Comment by tcharding** [comment]
I'm no project manager by a long shot, I looked at using GitHub projects but couldn't work out a good way. If anyone wants to suggest a better way to try and co-ordinate this I'm totally open to it.

**Reply by apoelstra** [reply]
No, this seems great to me!

**Reply by tcharding** [reply]
The only problem is @Kixunil and I can only edit the original post in this repo, not in bech32 and hex so the system might not work. Can we get edit perms without becoming maintainers?

**Reply by tcharding** [reply]
I did have a look at a repo I own but did not see any such setting (I couldn't find a repo that had collaborators though).

**Reply by apoelstra** [reply]
I can make you a maintainer of secp. For hex I think you already are one.

bech32 is special because we need to go through Clark anyway before making any decisions, so I think it's okay if it's harder to directly edit the issue.

**Reply by tcharding** [reply]
> I can make you a maintainer of secp.

ok, as long as that doesn't imply to anyone that I'm a cryptographer, which I'm not. I read a bunch of books on it but I didn't do a math major an uni.

**Comment by Kixunil** [comment]
I'm not really sure how much is this discussion is needed since I marked all 1.0 issues with a tag (feel free to add if some are missing). Maybe we should add those tags in other repos.

**Reply by apoelstra** [reply]
It's good to have a central place to coordinate 1.0 plans rather than debating every individual issue (and that's true even if there are issues filed for everything, including vague "we need to clean up X" ideas).

**Reply by Kixunil** [reply]
For things that are across all crates it makes more sense. Not sure about the individual ones.

**Reply by tcharding** [reply]
Yeah this was just the central discussion place. We could/should add '1.0' label to the other repos if we want to use that method all round. The crate specific issue was intended for crate specific discussion, it will be hard not to mention general things though but if all discussion is here it will be total chaos, long threads on github are impossible to follow.

**Reply by Kixunil** [reply]
Yes, please, let's use 1.0 everywhere. Tagging is much simpler than copy-pasting things. Also this may be a good use case for project board (but that one probably has to be synced manually anyway).

**Reply by tcharding** [reply]
I had a go at using a project board recently without much success. After reading [0] I think the missing piece is milestones.

[0] https://pavlenex.com/a-quick-guide-to-creating-a-roadmap-for-an-open-source-project-in-github

**Reply by Kixunil** [reply]
The trouble with milestones is there can be only one per issue. 1.0 issues are stabilization blockers but they don't imply deadline.

**Reply by tcharding** [reply]
Oh man, I'm super glad you pointed that out and I didn't have to discover it myself.

**Comment by Kixunil** [comment]
In the light of the news that BDK is trying to stabilize, here's an idea for stabilizing important types ("entities" from "Clean Architecture"; mainly `Block` and it's constituents, `Address`, `Network`...) much faster than the rest but with the cost of added churn and a bit of annoyance in downstream crates:

* Move them to separate crates as needed
* Remove any inherent methods we're not confident we could support long-term using the same API and move those to extension traits living in `bitcoin` or other unstable crate.
* Ensure their layout is what we want (I'm mainly thinking of tagging `Script`)
* If any important methods are implemented in traits and we're not confident about those traits make the methods inherent (I'm mainly thinking about hashes and `RawHash` stuff.)

Note that this includes **not** stabilizing seemingly important stuff like consensus encoding. However it's actually not that important for the goal of making it easier for crates to communicate via those types! The individual crates can still privately depend on unstable parts of the library while having stable API since each crate can have its own implementation. Example:

* crate `foo` exposes a method that returns `bitcoin-blockdata::Transaction` and internally decodes the transaction using `bitcoin:0.34`
* crate `bar` accepts `bitcoin-blockdata::Transaction` and encodes it using `bitcoin::0.35`
* `bar::consume_transaction(foo::create_transaction())` still works, each crate pulls its own version of `bitcoin`.

Of course this is somewhat ugly and worsens compile time but at least it's not completely broken and makes coordination between crates easier. And as stabilization continues this problem will decrease over time and eventually vanish.

**Reply by apoelstra** [reply]
This sounds good in principle, **but** "tagging `Script`" is a huge open-ended problem that needs to be solved before any of the other transaction and block types can bbe stabilized (since they all contain `Script`).

But I agree with this strategy to try to zero in on stabilizing the data types, and then branch out to the traits and stuff.

**Reply by Kixunil** [reply]
Do you have specific things you thing are uncertain about tagging script? It seems mostly straightforward to me.

**Reply by apoelstra** [reply]
@Kixunil depends how we want to tag them. If we want to tag scripts as being legacy/segwit/taproot then this is basically impossible because the `TxOut` type needs to have a `script_pubkey` field that can support any of them.

As a corollary we can't really make `Opcode` be dependent on taproot/segwit/legacy either, which kinda sucks.

But if we only want to tag scripts as scriptpubkey/scriptsig/witness-script/p2sh-embedded-script/scriptcode then I think this would be straightforward. We could use the type system to prevent these types of scripts from being confused with each other. Though I don't think we could do much more than that (e.g., we can't enforce length limits or opcode limits or whatever) since these limits depend on the legacy/segwit/taproot context.

**Reply by Kixunil** [reply]
If we do both then it's possible to allow casts, maybe even deref coercions (but that is probably wrong). So `LegacyPubkey` can be converted to `Pubkey` etc. `Opcode` can't be dependent on context but we can have multiple opcode types (yeah, maybe we don't want to).

**Reply by apoelstra** [reply]
@Kixunil I'd like that but I think we actually can't do it. Because a `Transaction` contains a `Vec<TxOut>` and each `TxOut` has a `Script`, and the different scripts may have different legacy/taproot/segwit tags.

So unless we wanted to hide the tag inside a `Box<dyn ScriptTag>` or something, which would largely defeat the point, we're stuck.

**Reply by Kixunil** [reply]
That's not a problem - the conversion is zero-cost, so when constructing transaction you'd just convert. If you already have a transaction it gets a bit messy converting back but I think it's still doable.

**Reply by apoelstra** [reply]
Convert to an untagged/generic scriptpubkey, you mean?

Then we'd basically have two `Script` regimes, one when the script is already in a transaction and one when the script is without, which have different type conversion rules. And which differently distinguish between taproot and segwit when serializing opcodes etc.

**Reply by Kixunil** [reply]
More like "less tagged". You still can't confuse public vs sig script vs tapscript vs witness program but can confuse script pubkeys. Although now that I re-read your message, is there anything to confuse? `script_pubkey` is not versioned or anything, so there's always only one.

**Reply by apoelstra** [reply]
@Kixunil :eyes: great observation that `script_pubkey` is not actually versioned and always uses the "bare" script interpreter. I guess the same is true of `script_sig`, of scripts directly inside witnesses, of scripts inside tapbranches inside witnesses, etc. Maybe there is no tagging needed beyond one that indicates the position inside the transaction.

**Reply by Kixunil** [reply]
If by "position inside the transaction" you mean also witness script and tapscript, not just `script_pubkey` and `script_sig`  then yes.

**Reply by apoelstra** [reply]
Yep. Witness scripts and tapscripts are encoded in distinct parts of the witness stack, so "position inside the transaction" still characterizes them.

Very cool.

**Reply by tcharding** [reply]
This is tracked in #1324
